File: php/test/AuthenticationEntityTest.php
<?php
declare(strict_types=1);

// Authentication entity test

require_once __DIR__ . '/../trexvpn_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AuthenticationEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = TRexVpnSDK::test(null, null);
        $ent = $testsdk->Authentication(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = authentication_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "authentication." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set TREXVPN_TEST_AUTHENTICATION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $authentication_ref01_ent = $client->Authentication(null);
        $authentication_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.authentication"), "authentication_ref01"));

        $authentication_ref01_data_result = $authentication_ref01_ent->create($authentication_ref01_data, null);
        $authentication_ref01_data = Helpers::to_map($authentication_ref01_data_result);
        $this->assertNotNull($authentication_ref01_data);

    }
}

function authentication_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/authentication/AuthenticationTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = TRexVpnSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["authentication01", "authentication02", "authentication03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("TREXVPN_TEST_AUTHENTICATION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "TREXVPN_TEST_AUTHENTICATION_ENTID" => $idmap,
        "TREXVPN_TEST_LIVE" => "FALSE",
        "TREXVPN_TEST_EXPLAIN" => "FALSE",
        "TREXVPN_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["TREXVPN_TEST_AUTHENTICATION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["TREXVPN_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["TREXVPN_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new TRexVpnSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["TREXVPN_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["TREXVPN_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: rb/test/authentication_entity_test.rb
# Authentication entity test

require "minitest/autorun"
require "json"
require_relative "../TRexVpn_sdk"
require_relative "runner"

class AuthenticationEntityTest < Minitest::Test
  def test_create_instance
    testsdk = TRexVpnSDK.test(nil, nil)
    ent = testsdk.Authentication(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = authentication_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "authentication." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set T_REX_VPN_TEST_AUTHENTICATION_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    authentication_ref01_ent = client.Authentication(nil)
    authentication_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.authentication"), "authentication_ref01"))

    authentication_ref01_data_result = authentication_ref01_ent.create(authentication_ref01_data, nil)
    authentication_ref01_data = Helpers.to_map(authentication_ref01_data_result.respond_to?(:data_get) ? authentication_ref01_data_result.data_get : authentication_ref01_data_result)
    assert !authentication_ref01_data.nil?
    assert !authentication_ref01_data["id"].nil?

  end
end

def authentication_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "authentication", "AuthenticationTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = TRexVpnSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["authentication01", "authentication02", "authentication03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["T_REX_VPN_TEST_AUTHENTICATION_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "T_REX_VPN_TEST_AUTHENTICATION_ENTID" => idmap,
    "T_REX_VPN_TEST_LIVE" => "FALSE",
    "T_REX_VPN_TEST_EXPLAIN" => "FALSE",
    "T_REX_VPN_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["T_REX_VPN_TEST_AUTHENTICATION_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["T_REX_VPN_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["T_REX_VPN_APIKEY"],
      },
      extra || {},
    ])
    client = TRexVpnSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["T_REX_VPN_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["T_REX_VPN_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
